tests/apollo/util/bft.py

```diff
@@ -256,21 +295,37 @@ def start_replica(self, replica_id):
         """
         if replica_id in self.procs:
             raise AlreadyRunningError(replica_id)
-        cmd = self.start_replica_cmd(replica_id)
-        self.procs[replica_id] = subprocess.Popen(cmd, close_fds=True)
 
-    def stop_replica(self, replica):
+        if self.is_existing:
+            subprocess.run(
+                self.start_replica_cmd(replica_id),
+                check=True
+            )
+            self.procs[replica_id] = self.replicas[replica_id]
```

Comment:
This is a bit confusing. It seems to imply that for existing replicas, the start_replica_cmd will automatically (and implicitly) fill in `self.replicas`.  I find this non-explicit behavior to be somewhat confusing. I'd prefer if instead we did the same thing we do for local procs and start the replica and explicitly get the value from start_replica_cmd. If that's not possible, because the replica is not actually the returned proc, which can certainly be possible since it's not local, I think we should instead have some wrapper function that gets run instead of calling `subprocess.run`. 

Take this suggestion with a grain of salt, as it may end up being more convoluted than what's going on, but at a minimum a comment would help. 

I bring all this up because I try to live by the mantra "Explicit is better than Implicit" when it comes to code.

Reply:
Thanks Andrew, I agree with you and will try to improve this. The tricky part is that in case of "external" replicas we don't have a direct representation of the process/container/whatever runtime. 
So in such cases self.procs[replica_id] is simply the Replica instance (effectively a stub representing the external runtime). 

I'll update the PR shortly.

Reply:
Thanks Teo! You rock!